simple_parsing/helpers/partial.py

```diff
@@ -0,0 +1,308 @@
+""" A Partial helper that can be used to add arguments for an arbitrary class or callable. """
+from __future__ import annotations
+
+import dataclasses
+import functools
+import inspect
+import typing
+from dataclasses import make_dataclass
+from functools import lru_cache, singledispatch, wraps
+from logging import getLogger as get_logger
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Hashable,
+    Sequence,
+    TypeVar,
+    _ProtocolMeta,
+    cast,
+    get_type_hints,
+    overload,
+)
+
+from typing_extensions import ParamSpec
+
+import simple_parsing
+
+__all__ = ["Partial", "config_dataclass_for", "infer_type_annotation_from_default"]
+
+C = TypeVar("C", bound=Callable)
+_P = ParamSpec("_P")
+_T = TypeVar("_T", bound="Any")
+_C = TypeVar("_C", bound=Callable[..., Any])
+
+logger = get_logger(__name__)
+
+
+@singledispatch
+def infer_type_annotation_from_default(default: Any) -> Any | type:
+    if isinstance(default, (int, str, float, bool)):
+        return type(default)
+    if isinstance(default, tuple):
+        return typing.Tuple[tuple(infer_type_annotation_from_default(d) for d in default)]
+    if isinstance(default, list):
+        if not default:
+            return list
+        # Assuming that all items have the same type.
+        return typing.List[infer_type_annotation_from_default(default[0])]
+    if isinstance(default, dict):
+        if not default:
+            return dict
+    raise NotImplementedError(
+        f"Don't know how to infer type annotation to use for default of {default}"
+    )
+
+
+@singledispatch
+def adjust_default(default: Any) -> Any:
+    """Used the adjust the default value of a parameter that we extract from the signature.
+
+    IF in some libraries, the signature has a special default value, that we shouldn't use as the
+    default, e.g. "MyLibrary.REQUIRED" or something, then a handler can be registered here to
+    convert it to something else.
+    """
+    return default
+
+
+@overload
+def cache_when_possible(fn: _C) -> _C:
```

Comment:
Personally, I'm not a huge fan of suffixes like `_when_possible` etc. They may be confusing, make discovery harder and are needlessly verbose IMO.
I think a caching mechanism implies that some values will be cached and others won't. The policy should be specified in documentation anyway.
So I'd prefer just `cache`.

Reply:
Woops. This is meant to be a module-private function. I'll rename it to `_cache_when_possible`.
And given that it's module-private, I'll will keep things as verbose as needed.